internal/kv/etcd/etcd_kv.go

```diff
@@ -64,6 +64,43 @@ func (kv *EtcdKV) GetPath(key string) string {
 	return path.Join(kv.rootPath, key)
 }
 
+func (kv *EtcdKV) WalkWithPagination(prefix string, paginationSize int, fn func([]byte, []byte) error) error {
+	start := time.Now()
+	prefix = path.Join(kv.rootPath, prefix)
+	ctx, cancel := context.WithTimeout(context.TODO(), RequestTimeout)
+	defer cancel()
+
+	batch := int64(paginationSize)
+	opts := []clientv3.OpOption{
+		clientv3.WithSort(clientv3.SortByKey, clientv3.SortAscend),
+		clientv3.WithLimit(batch),
+		clientv3.WithRange(clientv3.GetPrefixRangeEnd(prefix)),
+	}
+
+	key := prefix
+	for {
+		resp, err := kv.client.Get(ctx, key, opts...)
+		if err != nil {
+			return err
+		}
+
+		for _, kv := range resp.Kvs {
+			if err = fn(kv.Key, kv.Value); err != nil {
+				return err
+			}
+		}
+
+		if !resp.More {
+			break
+		}
+		// move to next key
+		key = string(append(resp.Kvs[len(resp.Kvs)-1].Key, 0))
```

Comment:
if prefix is key and here we get key1, can't just use key1 as prefix.

In fact the etcd implement prefix semantic using Get, it's not similar as RockDB scan

Reply:
With data A,AA, B
if you get(A) limit 3 you can not get B
if you scan start from A, you will only get A, AA.

In your implementation if data is A, AA, AB
then AB won't be selected out

Reply:
<img width="658" alt="image" src="https://user-images.githubusercontent.com/83447078/210212111-5ffc7fa9-cc2e-4770-bb08-960a5886073f.png">

That might be what we are looking for, range start, range end and a limit

Reply:
> With data A,AA, B
> if you get(A) limit 3 you can not get B
> if you scan start from A, you will only get A, AA.

This method just walks with a prefix, so can not get B is expected.

```
In your implementation if data is A, AA, AB
then AB won't be selected out
```
This implementation can select AB